builtin/builtin.go

```diff
@@ -309,6 +312,37 @@ func builtin_any(self, seq py.Object) (py.Object, error) {
 	return py.False, nil
 }
 
+const ascii_doc = `ascii(obj, /)
+Return an ASCII-only representation of an object.
+
+As repr(), return a string containing a printable representation of an
+object, but escape the non-ASCII characters in the string returned by
+repr() using \\x, \\u or \\U escapes. This generates a string similar
+to that returned by repr() in Python 2
+`
+
+func builtin_ascii(self, o py.Object) (py.Object, error) {
```

Comment:
I note that this has already been implemented **almost** here

https://github.com/go-python/gpython/blob/bd0985be3562d0a36eef5e34ecae4d4a4235fd57/py/string.go#L196

So you could factor that into its own function `stringEscape(s string, ascii bool)` with an extra parameter `ascii` which if set would encode everything, so wherever it says `strconv.IsPrint(c)` in that function you'd say `!ascii && strconv.IsPrint(c)`

That function deals with \r \t \n, the correct escaping for big chars and `'"'` vs `"'"`

Reply:
nice!

Reply:
Unfortunately, implementing `stringEscape(s string, ascii bool)` is not properly work in an easy case for this API.

for example
```
ascii('hello world')
# should be "'hello world'"
# but got '"\'hello world\'"'
```

Reply:
Note that if you are using the REPL then you want to print() the output otherwise it will be double escaped which looks like what has happened above.

I think using the `stringEscape` as suggested should fix the problems!  It will also fix quite a few corner cases you haven't considered yet!